8342303: Segfault in update_inherited_vtable: AppCDS, old bytecode, and redefineClasses

calvinccheung opened the pull request:
A simple fix to avoid referencing invalid vtable and itable entires in `InstanceKlass::restore_unshareable_info` if a class is not linked during CDS dump time.

Passed tiers 1 - 4 testing.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342303](https://bugs.openjdk.org/browse/JDK-8342303): Segfault in update_inherited_vtable: AppCDS, old bytecode, and redefineClasses (**Bug** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21667/head:pull/21667` \
`$ git checkout pull/21667`

Update a local copy of the PR: \
`$ git checkout pull/21667` \
`$ git pull https://git.openjdk.org/jdk.git pull/21667/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21667`

View PR using the GUI difftool: \
`$ git pr show -t 21667`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21667.diff">https://git.openjdk.org/jdk/pull/21667.diff</a>

</details>

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342303](https://bugs.openjdk.org/browse/JDK-8342303): Segfault in update_inherited_vtable: AppCDS, old bytecode, and redefineClasses (**Bug** - P4)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) Review applies to [d1801774](https://git.openjdk.org/jdk/pull/21667/files/d180177489299d45b7882b0ab267d9e9ffeb7a4a)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) Review applies to [d1801774](https://git.openjdk.org/jdk/pull/21667/files/d180177489299d45b7882b0ab267d9e9ffeb7a4a)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21667/head:pull/21667` \
`$ git checkout pull/21667`

Update a local copy of the PR: \
`$ git checkout pull/21667` \
`$ git pull https://git.openjdk.org/jdk.git pull/21667/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21667`

View PR using the GUI difftool: \
`$ git pr show -t 21667`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21667.diff">https://git.openjdk.org/jdk/pull/21667.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21667#issuecomment-2442699974)
</details>

calvinccheung commented:
/label remove hotspot

calvinccheung commented:
/label add hotspot-runtime

dholmes-ora commented on test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java:
```diff
@@ -0,0 +1,48 @@
+
+        System.out.println("Main: loading OldSuper");
+        // Load an old class (version 49), but not linking it.
+        Thread.currentThread().getContextClassLoader().loadClass("OldSuper");
```
Why the context classloader? Has it been set in this test?

dholmes-ora replied:
You kind of avoided the question though - why are you using the context classloader instead of e.g the classloader of the current class? Ther CCL is defined as follows:
```
Returns the context ClassLoader for this thread. The context ClassLoader may be set by the creator of the thread for use by code running in this thread when loading classes and resources. If not [set](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/Thread.html#setContextClassLoader(java.lang.ClassLoader)), the default is to inherit the context class loader from the parent thread.

The context ClassLoader of the primordial thread is typically set to the class loader used to load the application.
```
So if nothing sets it then you are just using the App/System loader, so why not just use that loader directly?

calvinccheung replied:
I've updated the tests to avoid using `getContextClassLoader()loadClass()`.
For the App/System loader case:
    `ClassLoader.getSystemClassLoader().loadClass()`
For the boot loader case:
    `Class.forName(<class name>, false, null)`

dholmes-ora reviewed (commented):
Functional fix looks good. Unclear about the the test details ...

Thanks

calvinccheung commented:
I've added two tests:

- redefine a super class loaded by the boot class loader;
- redefine an old super class loaded by the AppClassLoader.

Also simplified the OldClassAndRedefineClass.java test.

dholmes-ora commented on test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java:
```diff
@@ -0,0 +1,48 @@
+        // Redefine a class unrelated to the above old class.
+        Instrumentation instrumentation = InstrumentationRegisterClassFileTransformer.getInstrumentation();
+        System.out.println("INFO: instrumentation = " + instrumentation);
+        Class c = ClassLoader.getSystemClassLoader().loadClass("Hello");
```
```suggestion
        Class<?> c = ClassLoader.getSystemClassLoader().loadClass("Hello");
```

calvinccheung replied:
Fixed.

dholmes-ora commented on test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java:
```diff
@@ -0,0 +1,48 @@
+
+        System.out.println("Main: loading OldSuper");
+        // Load an old class (version 49), but not linking it.
+        ClassLoader.getSystemClassLoader().loadClass("OldSuper");
```
Why not `Class.forName` in all cases ??

calvinccheung replied:
I've changed it to use Class.forName().

dholmes-ora commented on test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineBootClassApp.java:
```diff
@@ -0,0 +1,69 @@
+    public static void main(String args[]) throws Throwable {
+        File bootJar = new File(args[0]);
+
+        Class superCls = Class.forName("BootSuper", false, null);
```
```suggestion
        Class<?> superCls = Class.forName("BootSuper", false, null);
```

calvinccheung replied:
Fixed.

dholmes-ora commented on test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineBootClassApp.java:
```diff
@@ -0,0 +1,69 @@
+            }
+        }
+
+        Class childCls = Class.forName("BootChild", false, null);
```
```suggestion
        Class<?> childCls = Class.forName("BootChild", false, null);
```

calvinccheung replied:
Fixed.

dholmes-ora commented on test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineOldSuperApp.java:
```diff
@@ -0,0 +1,69 @@
+    public static void main(String args[]) throws Throwable {
+        File bootJar = new File(args[0]);
+
+        Class superCls = ClassLoader.getSystemClassLoader().loadClass("OldSuper");
```
```suggestion
        Class<?> superCls = ClassLoader.getSystemClassLoader().loadClass("OldSuper");
```

calvinccheung replied:
Fixed.

dholmes-ora commented on test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineOldSuperApp.java:
```diff
@@ -0,0 +1,69 @@
+            }
+        }
+
+        Class childCls = ClassLoader.getSystemClassLoader().loadClass("NewChild");
```
```suggestion
        Class<?> childCls = ClassLoader.getSystemClassLoader().loadClass("NewChild");
```

calvinccheung replied:
Fixed.
Also, changed `ClassLoader.getSystemClassLoader().loadClass()` to `Class.forName()`.

dholmes-ora commented on test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java:
```diff
@@ -31,12 +31,12 @@ public static void main(String args[]) throws Throwable {
         Instrumentation instrumentation = InstrumentationRegisterClassFileTransformer.getInstrumentation();
         System.out.println("INFO: instrumentation = " + instrumentation);
-        Class c = ClassLoader.getSystemClassLoader().loadClass("Hello");
+        Class<?> c = ClassLoader.getSystemClassLoader().loadClass("Hello");
```
Can you use `Class.forName` here too.

calvinccheung replied:
Missed that one. Fixed now.

coleenp reviewed (commented):
There is a RedefineHelper class that hides all this instrumentation and agent code, and makes redefinition tests really simple.  I can't tell if you can use that from all the code in these tests. Examples are in the serviceability/jvmti/RedefineClasses tests.

calvinccheung commented:
> There is a RedefineHelper class that hides all this instrumentation and agent code, and makes redefinition tests really simple. I can't tell if you can use that from all the code in these tests. Examples are in the serviceability/jvmti/RedefineClasses tests.

In fact, some CDS tests are also using RedefineClassHelper. It did simply the test code. I've updated the tests.

dholmes-ora reviewed (approved):
Nothing further from me.

Thanks

iklam reviewed (approved):
LGTM

coleenp reviewed (approved):
Looks good.  A bit hard to follow since it uses classes already existing in the test directory, but that's good.

calvinccheung commented:
@iklam, @coleenp, @dholmes-ora,
I've noticed some failures from tier3 testing due to the jdk/test/lib/compiler/InMemoryJavaCompiler class not found. So I added the class to the "@build" command of the affected tests.
Please re-review. Probably one re-review is enough.
Thanks!

dholmes-ora reviewed (approved):
Okay on the updates

Thanks

calvinccheung commented:
Thanks for the reviews!

/integrate
